(WIP) reworking thumbnails handling

letsfindaway opened the pull request:
to avoid multiple and potentially heavy useless calls making the whole user experience very bad in terms of performances. Simplified messages that could also cause some lags.

This PR is for review only. It is not meant to be ever merged. Instead it helps me to review the changes and to comment on them, as I can mark files as visited, etc.

letsfindaway commented on src/gui/gui.pri:
```diff
@@ -4,7 +4,7 @@ HEADERS += src/gui/UBThumbnailView.h \
     src/gui/UBStylusPalette.h \
     src/gui/UBIconButton.h \
-    src/gui/UBThumbnailWidget.h \
+    src/gui/UBDocumentThumbnailsView.h \
```
Please also apply these changes to `gui/CMakeLists.txt`. It is straight forward and avoids having commits which do no longer build using `cmake`. I would very much like to avoid this situation.

kaamui replied:
We already had this discussion (and I made it clear when I merged cmake PRs), I'm not using cmake (at all) and am not able to test any changes I would make on the related files. I consider the cmake system community-driven, made for the Linux community by the Linux community, the cmake files are not even showing on Qt creator... so it doesn't come to my mind at all when working on OpenBoard. Moreover, I don't like to make any change that I can't even test a bit.

Of course I'll do it for this commit, now that you pointed it out and that it represents small changes where the possibility for me to introduce an error is small, but I'll probably do the error again until I use cmake myself, and I won't make the changes myself if tomorrow something I do in the project implies a lot of refactoring for these files

kaamui replied:
=> 0d11ea9fc8

I'll try to not forget it, especially for small changes like here, but won't promise it ^^

letsfindaway replied:
Thanks!

I understand your reasoning. But let me say: I also apply such changes to the `.pro` and `.pri` files, even if I don't use them. The probability for an error is veeery low.

OTOH I want to avoid that cmake breaks and I have to send PRs after such commits. So probably we can introduce some workflow for such changes: e.g. you don't push such things to `dev` or `master`, but use an intermediate working branch for that. Then I can even push the cmake changes to that branch. And the `dev` branch is never broken. Also commit squashing and fixups are possible, if you are not on `dev` or `master`.

letsfindaway replied:
I'm basing all my PRs on `dev`. Therefore yes, I assume that this is never broken. Also for others the branch to checkout if they want to see where the development goes is `dev`. So it would be good if this branch is at least "in a good condition".

letsfindaway reviewed (commented):
@kaamui: I created this PR on my project to better be able to review and comment on your changes. Here is my first feedback!

kaamui commented on src/adaptors/UBMetadataDcSubsetAdaptor.cpp:
```diff
@@ -88,7 +88,7 @@ void UBMetadataDcSubsetAdaptor::persist(std::shared_ptr<UBDocumentProxy> proxy)
     }
     QString fileName = proxy->persistencePath() + "/" + metadataFilename;
-    qWarning() << "Persisting document; path is" << fileName;
+    qWarning() << "Persisting document metadata; path is" << fileName;
```
I changed it a long time ago if I remember well, it was to have the information in the logs even in the release build in schools.

letsfindaway replied:
Still it seems to me that the reason was some debugging. And I don't know, but perhaps this debugging is now finished and resolved. So I think at least in the mid term this could be reduced to debug.

kaamui replied:
What about a `qInfo` ? I'd like to have the ability to verify how many times it is called (or if it not) in the next release if a teacher calls me to tell me that he has some bugs regarding the date of its file, or anything else.

letsfindaway replied:
A `qInfo` would be ok!

letsfindaway commented on src/board/UBBoardController.cpp:
```diff
@@ -2781,7 +2779,7 @@ void UBBoardController::freezeW3CWidget(QGraphicsItem *item, bool freeze)
+    //UBApplication::showMessage(tr("Reloading Board Thumbnails View (%1 pages)").arg(selectedDocument()->pageCount()));
     emit initThumbnailsRequired(selectedDocument());
-    UBApplication::showMessage(tr("Board Thumbnails View up-to-date"));
+    //UBApplication::showMessage(tr("Board Thumbnails View up-to-date"));
```
So you assume that you can avoid `reloadThumbnails` completely and just do the updates of exactly the affected ones? That would be great!

kaamui replied:
I'm trying to !! But have some weird bug on the QGraphicsView ^^

kaamui replied:
Because of the hurry I'm in though, there are some cases where I'll let the reload being the way the thumbnails are handled, to make things simpler. But during the analysis, I also found that a lot of useless operations are made on the labels in Document Mode !  `setGraphicsItems` is not well designed

letsfindaway commented:
With the current `dev` as of `1.7.2rc-241004` I have an issue when resizing the left sidebar.

When I e.g. increase the size of the sidebar, then the icons get larger. That's ok. But if you scroll down to the bottom, not all pages are visible. OTOH, if you make the sidebar smaller and the icons get smaller, then there is an empty area at the end.

It seems as if the height of the scroll area is not adjusted according to the overall size of the thumbnails.

letsfindaway commented:
Another bug: Metadata is not always written. To reproduce:

- Start OpenBoard
- Do some simple drawing or whatever else to the scene
- Close OpenBoard using the menu or Alt+F4
- The resulting document does not have a `metadata.rdf` file
- When restarting OpenBoard, the document is therefore not loaded in Document view and cannot be opened

kaamui commented:
ouch ! thank you for finding this before end users do !

kaamui commented:
I didn't expect that the order of creation of metadata or pages would matter. It should probably not.

kaamui commented:
Also, I wonder if we should not create them (metadata.rdf) when we find them to be missing. Do you see any reason why we should/could not ?

letsfindaway commented:
> Also, I wonder if we should not create them (metadata.rdf) when we find them to be missing. Do you see any reason why we should/could not ?

Can we derive all necessary information from the other files? But I think we can:

```xml
<RDF xmlns="http://www.w3.org/1999/02/22-rdf-syntax-ns#" xmlns:dc="http://purl.org/dc/elements/1.1/" xmlns:ub="http://uniboard.mnemis.com/document"> 
    <!-- just create an arbitrary UUID for the document -->
    <Description about="http://uniboard.mnemis.com/document/df912da6-1fcb-4f43-8c6d-2aa6caf1d728">
        <!-- Folder creation date in another format -->
        <dc:title>08.09.24 11:16</dc:title>
        <!-- type seems to contain documentGroupName. Leave it empty -->
        <dc:type></dc:type>
        <!-- again date of the folder -->
        <dc:date>2024-09-08T09:16:26Z</dc:date>
        <!-- contstant -->
        <dc:format>image/svg+xml</dc:format>
        <!-- same UUID as above -->
        <dc:identifier>http://uniboard.mnemis.com/document/df912da6-1fcb-4f43-8c6d-2aa6caf1d728</dc:identifier>
        <!-- constant -->
        <ub:version>4.8.0</ub:version>
        <!-- from UBSettings::settings()->pageSize? -->
        <ub:size>1696x960</ub:size>
        <!-- same date as above or newest date of any page file in this folder -->
        <ub:updated-at>2024-09-08T09:17:13Z</ub:updated-at>
    </Description>
</RDF>
```

But I hope that documents with missing metadata are very rare, so this does not seem to be important to me.

letsfindaway commented:
Note that metadata is still not updated when a scene is written. I would expect that `ub:updated-at` would reflect the latest update of a scene. This is not the case.

To minimize writing metadata one could also think about introducing a "dirty" flag and only actually save the metadata when switching documents, closing OpenBoard or after some timeout (like with the settings).

letsfindaway commented:
I'm closing this PR now as I think that the topic it was about is now solved and `dev` has again a stable version.
